### PR TITLE
GH-121970: Rewrite the C-API annotations extension

### PR DESCRIPTION
**Pull request by AA-Turner**
I've tested the output before and after this commit, entirely equal save for a new link target in ``c-api/module.html#c.Py_mod_create.create_module`` and ``c-api/module.html#c.Py_mod_exec.exec_module`` (because we switch from patching in ``:noindex:`` to using the C Domain's supported ``:no-index-entry:``).

This introduces some new dataclasses, removes the state stored in closures and the ``Annotated`` class instance, and does some small cleanups. Notably, ``stable_abi.dat`` is changed to storing the *role* name rather than the *object type*, which simplifies things a little.

A

----
📚 Documentation preview 📚: https://cpython-previews--121985.org.readthedocs.build/en/121985/c-api/index.html

* Issue: gh-121970

### CODE REVIEW

File: Doc/tools/extensions/c_annotations.py [hunk 1]
```diff
@@ -1,226 +1,281 @@
...
+    #: Return type of the function.
+    result_type: str = ''
+    #: Reference count effect for the return value.
+    result_refs: int | None = None
+
+
+@dataclasses.dataclass(frozen=True, slots=True)
+class StableABIEntry:
+    #: Role of the object.
+    #: Source: Each [item_kind] in stable_abi.toml is mapped to a C Domain role.
```
Comment (encukou) [inline Doc/tools/extensions/c_annotations.py]: What does the `#:` mean?

If the attribute is named `role`, the comment `Role of the object` seems entirely unnecessary to me.

File: Doc/tools/extensions/c_annotations.py [hunk 2]
```diff
@@ -1,226 +1,281 @@
...
-def init_annotations(app):
-    annotations = Annotations(
-        path.join(app.srcdir, app.config.refcount_file),
-        path.join(app.srcdir, app.config.stable_abi_file),
+@dataclasses.dataclass(slots=True)
+class RefCountEntry:
+    #: Name of the function.
+    name: str
+    #: List of (argument name, type, refcount effect) tuples.
+    args: list[tuple[str, str, int | None]] = dataclasses.field(default_factory=list)
```
Comment (encukou) [inline Doc/tools/extensions/c_annotations.py]:
```suggestion
    #: List of (argument name, type, refcount effect) tuples.
    # (Currently not used. If it was, a dataclass might work better.)
    args: list = dataclasses.field(default_factory=list)
```

**Review by encukou — COMMENTED** [review]
Looks good, thanks for refreshing this extension!

Some nitpicks:

File: Doc/tools/extensions/c_annotations.py (hunk 1 above)
Comment (AA-Turner) [reply Doc/tools/extensions/c_annotations.py]: It's a [Sphinx-style docstring-comment](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute), but I can change this to an attribute docstring if you'd prefer?

Comment (encukou) [reply Doc/tools/extensions/c_annotations.py]: It's not documented, and wouldn't be documented even if this was a public tool, so I think a regular comment will do.
I don't think we can expect others to know about these Sphinx features, so they won't be maintained as valid docstring-comments or attribute-docstrings.

But, thinking more about the target audience here: Could you add *"See `StableABIEntry` in `Doc/tools/extensions/c_annotations.py` for descriptions of the fields"* to `gen_doc_annotations` in `Tools/build/stable_abi.py`?

Comment (AA-Turner) [reply Doc/tools/extensions/c_annotations.py]: Done.

**Review by encukou — APPROVED** [review]

**Pull request merged**